### DISCUSSION TITLE
Annoted types as body leads to AssertionError

### DISCUSSION THREAD

**Discussion by orfisko** [fastapi/fastapi#9898] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Literal, Annotated, Union

from fastapi import FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel, Field


class Foo(BaseModel):
    classType: Literal["Foo"] = "Foo"
    foo_field: str


class Bar(BaseModel):
    classType: Literal["Bar"] = "Bar"
    bar_field: str


FooBar = Annotated[Union[Foo, Bar], Field(discriminator="classType")]

app = FastAPI()

@app.post("/")
async def post(
        foo_bar: FooBar,
):
    print(foo_bar)


testclient = TestClient(app)
foo = Foo(foo_field="foo")
testclient.post(
    url="/",
    data=foo.model_dump_json()
)
# AssertionError: Param: foo_bar can only be a request body, using Body()
```

### Description

With pydantic v2 I hoped we could end the problems with unions in earlier versions. So, I went along and tried to set something up. Using a discriminator is a fool proof way to make sure the union will actually work in Pydantic, so that's what I defined. When using this as a body parameter though, fastapi is not able to cope with it it seems.

### Operating System

Linux

### FastAPI Version

0.100.0

### Python Version

3.10.6

**Comment by Kludex** [comment, marked as answer]
FastAPI doesn't support `pydantic.Field` annotation metadata yet.

```py
from typing import Literal, Annotated, Union

from fastapi import FastAPI, Body
from fastapi.testclient import TestClient
from pydantic import BaseModel


class Foo(BaseModel):
    classType: Literal["Foo"] = "Foo"
    foo_field: str


class Bar(BaseModel):
    classType: Literal["Bar"] = "Bar"
    bar_field: str


FooBar = Annotated[Union[Foo, Bar], Body(discriminator="classType")]  # Change here.

app = FastAPI()


@app.post("/")
async def post(foo_bar: FooBar):
    print(foo_bar)


testclient = TestClient(app)
foo = Foo(foo_field="foo")
testclient.post(url="/", data=foo.model_dump_json())
testclient.post(url="/", data=Bar(bar_field="bar").model_dump_json())
```

**Reply by vianmixtkz** [reply]
Hi @Kludex ,

I used to define several children classes of pydantic `ConstrainedStr` mainly to  validate my strings against some regex.
I could then easily use those classes in my FastAPI app, as path of query parameters and FastAPI would understand automatically

Now that I moved to pydantic V2, I converted those classes to `Annotated[str, Field(...)]` but now FastAPI does not understand and I have to re-define my type for every use case.

For example,  let's say I define:

`SearchString = Annotated[str, Field(min_length=3)]`

I then have to define:

`SearchStringQuery = Annoated[str, Query(min_lenght=3)]`

Are there plans to improve FastAPI comptatibility with pydantic V2 ?

**Reply by tiangolo** [reply]
Yes, there are plans to support it. 🚀
